crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -219,10 +212,18 @@ impl<'a> WasiSnapshotPreview1 for WasiCtx {
         if offset > i64::max_value() as u64 {
             return Err(Errno::Io);
         }
-        let host_nread = entry
-            .as_handle(&required_rights)?
-            .preadv(&mut buf, offset)?
-            .try_into()?;
+
+        let host_nread = {
+            let mut buf = guest_slices
+                .iter_mut()
+                .map(|s| io::IoSliceMut::new(&mut *s))
```

Comment:
So the cool thing about the previous approach was that we didn't have to do iterate twice over the same array, which we unfortunately do now. I haven't look in-depth into `GuestSlice`, so I'm not sure how to go about managing lifetimes here yet, but would it be possible to store `io::IoSliceMut` in the first pass rather than do the conversion from `GuestSlice` to `&mut [u8]` here?

Reply:
I think `guest_slices` could presumably be `Vec<IoSliceMut>` here, right? I don't think there's any intrinsic reason a second iteration is required?

Reply:
Yep, that'd be ideal.

Reply:
This is the trickiest thing I had to deal with with these changes. For safety, we need the `GuestSlice` to stay alive for at least as long as any `&mut [u8]` we get from it (Deref trait enforces this). I couldn't find any other mechanism to collect all of the `GuestSlice`s (creating them does the borrow check) and then use them as `&mut [u8]` without iterating through them twice. I agree its not great, but the number of vectors hopefully won't be big enough that this becomes a big performance hit. If it does become a problem, we could perhaps find a faster mechanism that involves some manual `unsafe`, but I'd prefer not to cross that bridge yet.